AboutStep: Improve site topic suggestions

jsnmoon opened the pull request:
This PR improves the suggestion logic for the site topic input to ~match the beginning and end of words~ prioritize suggestions beginning with the user's query. It also creates an A/B test named `aboutSuggestionMatches` starting on `20180704`. We hypothesize that this will increase the quality of site topic selections.

<img width="521" alt="create_a_site_ _wordpress_com" src="https://user-images.githubusercontent.com/4044428/42002701-6e03099c-7a25-11e8-8a67-43522e56b633.png">

This PR also fixes a bug relating to the [About step of the signup flow](http://calypso.localhost:3000/start/about) where we invoke `setState` on an unmounted component.

# Testing instructions
1. Spin up this branch locally.
2. Navigate to the [signup flow](http://calypso.localhost:3000/start/about).
3. Using the A/B test helper component on the bottom right of the screen, set `aboutSuggestionMatches` test to `beginningAndEndOfWords`. (EDIT: Try `enhancedSort` now instead of `beginningAndEndOfWords`.)

<img width="505" alt="create_a_site_ _wordpress_com" src="https://user-images.githubusercontent.com/4044428/42001675-ef66fcfe-7a21-11e8-98fa-1f9c07c5fb20.png">

4. Try a variety of queries for the site topic input. Ensure that they only match the beginning or ending of words.

bluefuton commented:
@jsnmoon have you got any examples of search phrases that didn't work so well before these changes were made?

sirino commented:
@bluefuton one example was when you start to search for `travel` and the first match that comes up after typing the first three letters `tra` is `heat transfer vinyl`.

jsnmoon commented:
> one example was when you start to search for `travel` and the first match that comes up after typing the first three letters `tra` is `heat transfer vinyl`.

> have you got any examples of search phrases that didn't work so well before these changes were made?

To be honest, I don't think this change does enough to enhance our users' experiences. Take the `tra` query, for example: Even with this PR, `heat transfer vinyl` is the first suggestion recommended to the user. It'd make more sense to weight suggestions starting with the query more than suggestions containing `tra` later in the middle of the suggestion.

I'll see if there's a simple, straightforward way of doing this.

jsnmoon commented:
I updated this PR to sort beginning matches higher. I also reverted the filtering logic change. 

What do you think? See examples for query `tra` below.

Before:

<img width="477" alt="create_a_site_ _wordpress_com" src="https://user-images.githubusercontent.com/4044428/42116915-fa415dcc-7bb6-11e8-85c1-f97bdbee0be8.png">


After:
<img width="475" alt="create_a_site_ _wordpress_com" src="https://user-images.githubusercontent.com/4044428/42116903-ec4ca186-7bb6-11e8-9037-d4108451a5c1.png">

dmsnell commented on client/signup/steps/about/index.jsx:
```diff
@@ -93,14 +98,13 @@ class AboutStep extends Component {
-		this.setState( { siteTopicValue: event.target.value } );
+	handleSuggestionChangeEvent = ( { target: { name, value } } ) => {
+		this.setState( { query: value } );
+		this.setState( { siteTopicValue: value } );
```
is there a reason we aren't setting these together? or of sequencing the following updates until after setting the state?

```js
handleSuggestionChangeEvent = ( { target: { name, value } } ) => this.setState( {
	query: value,
	siteTopicValue: value,
}, () => {
	this.props.recordTracksEvent( 'calypso_signup_actions_select_site_topic', { value } ),
	this.formStateController.handleFieldChange( { name, value } );
} );
```

jsnmoon replied:
In hindsight, your suggested change would be much better. I'll look into putting this change in when we rip out the A/B test.

dmsnell commented:
Have we considered turning this into an API call? What is the value proposition of having these suggestions inside of Calypso? We don't do anything with them do we? If we had control on the server we could adjust the algorithm as we see fit.

jsnmoon commented:
> Have we considered turning this into an API call? What is the value proposition of having these suggestions inside of Calypso? [...] If we had control on the server we could adjust the algorithm as we see fit. (@dmsnell)

👍 I think making this an API call would be good. 

> We don't do anything with them do we? (@dmsnell)

@Automattic/cobalt eventually plans to use these results for recommending domains and subdomains in the domain step.

dmsnell commented:
> eventually plans to use these results

okay. seems strange to introduce it early on but not use it. (p.s. until there's code in the repo it will _never_ be used 😉 )

if we're planning on using it maybe we should just rip it all out for now until we need it?
